### PR TITLE
Add documentation for mpl_toolkits.axes_grid1.inset_locator

### PR DESCRIPTION
**Pull request by sargas**
This module has some nice, undocumented (except in narrative form) helpers that has been interesting to figure out.

I made some formatting changes, and some slight API differences:

1. `AnchoredSizeLocator` and `AnchoredZoomLocator` both overrode `__call__` merely to save the axes (which is unused in `AnchoredSizeLocator` anyways). I moved the saving of this value to the `__call__` of their superclass. No other file in matplotlib uses these classes, and the superclass doesn't look used for anything else online.
2. `inset_axes` and `zoomed_inset_axes` take a `**kwarg`, but the only keyworded argument that would not result in an error is `borderpad`. I changed the functions to make that clear.

I still have `AnchoredLocatorBase` (and its subclasses) to document, but I don't understand the code as is - for instance, a required parameter `offsetbox` is always unused, and I don't understand why a locator object (something used with `Axes.set_axes_locator`) would ever have any method except those used by `__call__` (making the capabilities inherited from `AnchoredOffsetBox` useless).

### CODE REVIEW

**Comment by WeatherGod** [comment]
Docbuild failure, but I don't know what it means: https://api.travis-ci.org/jobs/74123925/log.txt?deansi=true

**Comment by sargas** [comment]
I generated the docs in py3 on my local machine, it turns out the `dedent_interpd` decorator doesn't like to work on classes in py2.

I moved the class-level docstrings to the `__init__` functions; that allows the docs to build here.
I don't know how the python3.3 error (in `matplotlib.tests.test_backend_ps.test_savefig_to_stringio_with_usetex`) could have happened though.

**Comment by WeatherGod** [comment]
Ok, that makes sense. Don't worry about that usetex test, it is a transient
one that has to do with the Travis system.

**Comment by tacaswell** [comment]
Awesome!

I have no problem with this sneaking in under the wire for 1.5, but do not have time to review.

**Comment by WeatherGod** [comment]
A question for the other devs... Is there any reason why we continue to have the axes_grid and axes_grid1? From my understanding, axes_grid is "deprecated" and is only kept around for compatibility. Documentation-wise, it looks like only the axes_grid module is actually generated.

**Comment by jenshnielsen** [comment]
I would be :+1: on removing axes_grid

**Comment by tacaswell** [comment]
Has anyone been able to review this?

**Comment by tacaswell** [comment]
@leejjoon Can you review this?

File: lib/mpl_toolkits/axes_grid1/inset_locator.py [hunk 1]
```diff
@@ -1,50 +1,77 @@
+"""
+A collection of functions and objects for creating or placing inset axes.
+"""
 from __future__ import (absolute_import, division, print_function,
                         unicode_literals)
 
-from matplotlib.externals import six
```
Comment (tacaswell) [inline lib/mpl_toolkits/axes_grid1/inset_locator.py]: please leave the six import

File: lib/mpl_toolkits/axes_grid1/inset_locator.py [hunk 2]
```diff
@@ -1,50 +1,77 @@
...
+        super(AnchoredLocatorBase, self).__init__(
+            loc, pad=0., child=None, borderpad=borderpad,
+            bbox_to_anchor=bbox_to_anchor, bbox_transform=bbox_transform
+        )
 
     def draw(self, renderer):
         raise RuntimeError("No draw method should be called")
 
     def __call__(self, ax, renderer):
+        self.axes = ax
```
Comment (tacaswell) [inline lib/mpl_toolkits/axes_grid1/inset_locator.py]: what is this for?

File: lib/mpl_toolkits/axes_grid1/inset_locator.py [hunk 3]
```diff
@@ -129,13 +145,21 @@ def get_extent(self, renderer):
...
     def __init__(self, bbox, **kwargs):
+        """
+        Patch showing the shape bounded by a Bbox.
+
+        Parameters
+        ----------
+        bbox : `matplotlib.transforms.Bbox`
+            Bbox to use for the extents of this patch.
+
+        The kwargs are Patch properties:
```
Comment (tacaswell) [inline lib/mpl_toolkits/axes_grid1/inset_locator.py]: Can you put this in a `Notes` section to keep numpydoc happy?

File: lib/mpl_toolkits/axes_grid1/inset_locator.py [hunk 4]
```diff
@@ -196,24 +271,39 @@ def connect_bbox(bbox1, bbox2, loc1, loc2=None):
...
+        loc2 : {1, 2, 3, 4}, optional
+            Corner of *bbox2* to draw the line. If None, defaults to *loc1*.
+            Valid values are::
+
+                'upper right'  : 1,
+                'upper left'   : 2,
+                'lower left'   : 3,
+                'lower right'  : 4
+
+        The kwargs are Patch properties for the line drawn:
```
Comment (tacaswell) [inline lib/mpl_toolkits/axes_grid1/inset_locator.py]: Same with the `Notes` section here

File: lib/mpl_toolkits/axes_grid1/inset_locator.py [hunk 5]
```diff
@@ -276,7 +456,7 @@ def inset_axes(parent_axes, width, height, loc=1,
                                        width, height,
                                        loc=loc,
                                        bbox_transform=bbox_transform,
-                                       **kwargs)
+                                       borderpad=borderpad)
```
Comment (tacaswell) [inline lib/mpl_toolkits/axes_grid1/inset_locator.py]: Just checking that this fully exhausts all kwargs that can be passed to `AnchoredSizeLocator`?

**Comment by tacaswell** [comment]
Gave this a quick read, looks really good :+1: 

There are a bunch of places where things need to be put under a `Notes` section for numpydoc to render them nicely.

In the places where `**kwargs` is eliminated, it should be checked that all possible keyword args really are covered.

File: lib/mpl_toolkits/axes_grid1/inset_locator.py (hunk 2 above)
Comment (sargas) [inline lib/mpl_toolkits/axes_grid1/inset_locator.py]: Both `AnchoredZoomLocator` and `AnchoredSizeLocator` save the axes in `__call__` for use in `get_extent`. This was just an attempt at removing code duplication, although admittingly it isn't documented clearly for subclasses.

Comment (tacaswell) [inline lib/mpl_toolkits/axes_grid1/inset_locator.py]: fair enough.

**Comment by sargas** [comment]
Thanks for reviewing this.

I don't think the `kwargs` should go into `Notes`, since the keyworded arguments are parameters.

I attempted to do the following in `Parameters`:
```
    kwargs
        Patch properties for the lines and box drawn:
        %(Patch)s
```
This doesn't quite get the formatting right either (the start of the description is misaligned with the start of the descriptions of the other parameters). The generated HTML has an extra `<blockquote>` element causing this.

**Comment by tacaswell** [comment]
I think the priority should be that the docs render _reasonably_ even if they are not semantically correct.

**Comment by sargas** [comment]
Okay. I've also double checked the **kwargs options I removed.

**Comment by tacaswell** [comment]
This is what the built docs look like http://matplotlib.org/devdocs/mpl_toolkits/axes_grid/api/inset_locator_api.html

**Comment by tacaswell** [comment]
back-ported to 1.5.x as 8677af6

**Pull request merged**